src/libproc_macro/bridge.rs

```diff
@@ -0,0 +1,856 @@
+                data: BTreeMap::new(),
+            }
+        }
+    }
+
+    impl<T> Owned<T> {
+        pub(super) fn alloc(&mut self, x: T) -> Handle {
+            let counter = self.counter.fetch_add(1, Ordering::SeqCst);
+            let handle = Handle::new(counter as u32)
+                .expect("`proc_macro` handle counter overflowed");
+            assert!(self.data.insert(handle, x).is_none());
+            handle
+        }
+
+        pub(super) fn take(&mut self, h: Handle) -> T {
+            self.data.remove(&h).expect("use-after-free in `proc_macro` handle")
+        }
+    }
+
+    impl<T> Index<Handle> for Owned<T> {
+        type Output = T;
+        fn index(&self, h: Handle) -> &T {
+            self.data.get(&h).expect("use-after-free in `proc_macro` handle")
+        }
+    }
+
+    impl<T> IndexMut<Handle> for Owned<T> {
+        fn index_mut(&mut self, h: Handle) -> &mut T {
+            self.data.get_mut(&h).expect("use-after-free in `proc_macro` handle")
+        }
+    }
+
+    pub(super) struct Interned<T: 'static> {
+        owned: Owned<T>,
+        interner: HashMap<T, Handle>,
+    }
+
+    impl<T: Copy + Eq + Hash> Interned<T> {
+        pub(super) fn new(counter: &'static AtomicUsize) -> Self {
+            Interned {
+                owned: Owned::new(counter),
+                interner: HashMap::new(),
+            }
+        }
+
+        pub(super) fn alloc(&mut self, x: T) -> Handle {
+            let owned = &mut self.owned;
+            *self.interner.entry(x).or_insert_with(|| owned.alloc(x))
+        }
+
+        pub(super) fn take(&mut self, h: Handle) -> T {
+            self.owned[h]
+        }
+    }
+}
+
+macro_rules! define_handles {
+    (
+        'owned: $($oty:ident,)*
+        'interned: $($ity:ident,)*
+    ) => {
+        #[repr(C)]
+        #[allow(non_snake_case)]
+        struct Counters {
+            $($oty: AtomicUsize,)*
+            $($ity: AtomicUsize,)*
+        }
+
+        static COUNTERS: Counters = Counters {
+            $($oty: AtomicUsize::new(1),)*
+            $($ity: AtomicUsize::new(1),)*
+        };
+
+        #[repr(C)]
+        #[allow(non_snake_case)]
+        struct Storage<S: server::Types> {
+            $($oty: storage::Owned<S::$oty>,)*
+            $($ity: storage::Interned<S::$ity>,)*
+        }
+
+        impl<S: server::Types> Storage<S> {
+            fn new(counters: &'static Counters) -> Self {
+                Storage {
+                    $($oty: storage::Owned::new(&counters.$oty),)*
+                    $($ity: storage::Interned::new(&counters.$ity),)*
+                }
+            }
+        }
+
+        $(
+            #[repr(C)]
+            pub(crate) struct $oty(storage::Handle);
+            impl !Send for $oty {}
+            impl !Sync for $oty {}
+
+            // Forward `Drop::drop` to the inherent `drop` method.
+            impl Drop for $oty {
+                fn drop(&mut self) {
+                    $oty(self.0).drop();
```

Comment:
Doesn't this cause a loop? Wouldn't `$oty(self.0)` drop again right after we call the inherent `drop`?

Reply:
The inherent `drop` takes `self` by value and results in the temporary handle's associated data being taken from storage, and the handle passed to `mem::forget`, so the temporary handle never gets dropped.